APMrover2/sailboat.cpp

```diff
@@ -127,111 +171,178 @@ void Sailboat::update_mainsail(float desired_speed)
     // use PID controller to sheet out
     const float pid_offset = rover.g2.attitude_control.get_sail_out_from_heel(radians(sail_heel_angle_max), rover.G_Dt) * 100.0f;
 
-    mainsail = constrain_float((mainsail+pid_offset), 0.0f ,100.0f);
+    // See if we should sheet out to slow down
+    float speed_offset = 0.0f;
+    if (!is_positive(throttle_out) && enable == 3) {
+        speed_offset = throttle_out;
+        // never try to reverse against the sail
+        throttle_out = 0.0f;
+    }
+
+    mainsail = constrain_float((mainsail + pid_offset + speed_offset), 0.0f ,100.0f);
     rover.g2.motors.set_mainsail(mainsail);
+
+    // see if we should allow throttle
+    if (throttle_state == sailboat_throttle::NEVER ||
+       !throttle_assist()) {
+        throttle_out = 0.0f;
+        if (enable == 2) {
+            rover.g2.attitude_control.relax_throttle();
+        }
+    }
+
+    // see if we should limit the throttle allowed
+    if ( (sail_options & options::AUX_THROTTLE) != 0 && (sail_options & options::AUX_THROTTLE_LIMIT) != 0 ) {
+        RC_Channel *rc_ptr = rc().find_channel_for_option(RC_Channel::AUX_FUNC::SAILBOAT_THR);
```

Comment:
I'm a little concerned that this breaks the pattern we have in Copter and Rover re consumption of the pilot input.  We normally consume the pilot's input at the top of the flight mode code.  This unfortunately buries is two levels below the flight code.  When we break this pattern it makes it much more difficult to figure out what is going on.

Reply:
I think we should probably use the regular throttle channel for throttle and then allow the pilot to control the main sail with channel 4.

Reply:
I have removed this throttle limit stuff, not sure how usefully it would be in anycase